### PR TITLE
feat(broker-core): add debug exporter http server

### LINKED ISSUE #1550: I can inspect exported events using a web browser
The intention is to provide a developer friendly way to debug the record stream. Which can also be used to explain the behavior of Zeebe in discussions.

AT:
- I can specify a port for the debug exporter, if the port is > 0 a web server is started which exposes the record stream
- the records are ordered by position descending
- the list of records shows the position, key, value type, record type and intent
- I can expand a record value which shows the whole record as JSON

### PR DESCRIPTION
**Pull request by menski**
- adds an optional http server to the debug exporter
- if debug exporter is configured with a port greater then zero the http server will be enabled
- the exporter then serves a html page of the latest records
- the records can be filtered by value type and full text search
- the maximum number of records served can be configured `httpLimit`

closes #1550

### Screenshot

screenshot-2018-10-25-07 55 29

### CODE REVIEW

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+    switch (record.metadata.valueType) {
+      case 'WORKFLOW_INSTANCE':
+        return $('#filter-workflow-instance');
+      case 'JOB':
+        return $('#filter-job');
+      case 'INCIDENT':
+        return $('#filter-incident');
+      case 'TIMER':
+        return $('#filter-timer');
+      case 'MESSAGE_SUBSCRIPTION':
```
Comment (saig0) [inline broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: The value type `MESSAGE` should be also displayed if `Message` is selected.

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+      <th scope="col">BPMN Process Id</th>
+      <th scope="col">Element Id</th>
+      <th scope="col">Workflow Instance Key</th>
+      <th scope="col">Scope Instance Key</th>
+      <th scope="col">Element Instance Key</th>
+    </tr>
+  </thead>
+  <tbody>
+  {{ #. }}
+    <tr data-toggle="collapse" data-target="#collapse-{{_id}}" aria-expanded="false" aria-controls="collapse-{{_id}}">
```
Comment (saig0) [inline broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: It should be more clear that I can click on the record to see the details. I found this feature by a fluke ;)

**Review by saig0 — CHANGES_REQUESTED** [review]
It works and is kind of nice :+1:  However, I'm not sure if the debug exporter is the right place to put it in and who need this :sweat_smile:

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporter.java [hunk 3]
```diff
@@ -37,6 +40,11 @@ public void configure(Context context) {
     log = context.getLogger();
     configuration = context.getConfiguration().instantiate(DebugExporterConfiguration.class);
     logLevel = configuration.getLogLevel();
+
+    if (httpServer == null && configuration.httpPort > 0) {
+      httpServer =
+          new DebugExporterHttpServer(log, configuration.httpPort, configuration.httpLimit);
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporter.java]: I wanted to start it only once, i.e. not for every partition, and verify that the port is available, which I think should happen in the initial configure call. But I can also move it to open if it is to strange.

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+  }
+
+  public void close() {
+    if (server != null) {
+      server.stop(0);
+      server = null;
+    }
+  }
+
+  public synchronized void add(String recordJson) {
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java]: Cause multiple exporters, i.e. for every partition, will add records to the list.

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      httpServer.start();
+      log.info("Debug exporter http server started, see http://localhost:" + port);
+      return server;
+    } catch (IOException e) {
+      log.warn("Failed to start debug exporter http server", e);
+    }
+    return null;
+  }
+
+  private byte[] loadIndexPage() {
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java]: Okay I can change that, but then I think I have to open the server on the initial configure call. At least I understood it that way. @npepinpe what do you think?

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html [hunk 6]
```diff
@@ -0,0 +1,185 @@
+<!doctype html>
+<html lang="en">
+<head>
+  <meta charset="utf-8">
+  <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
+  <title>Zeebe Debug Exporter</title>
+  <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.1.3/css/bootstrap.min.css" integrity="sha384-MCw98/SFnGE8fJT3GXwEOngsV7Zt27NXFoaoApmYm81iuXoPkFOJwJ8ERdknLPMO" crossorigin="anonymous">
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: I didn't wanted to implement a static resource handler in the web server and keep it simple. But if you want I can also serve all resource with the HTTP server

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+            r._elementInstanceKey = r.value.headers.activityInstanceKey;
+            break;
+      }
+      return r;
+    });
+  }
+
+  function searchRecords(records) {
+    var searchTerm = $('#filter-search').val();
+    console.log(searchTerm);
```
[the remark this answers is not in the document]
Comment (menski) [reply broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: oops forgot to remove the debugging code

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html (hunk 1 above)
Comment (menski) [reply broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: Correct I forgot it, will change

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (npepinpe) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java]: If you fail in the configure call, then the broker will not start; if that's the intention then you'd have to throw an exception in the configure call (I'm still not 100% convinced it wouldn't have been better to have a separate `validate` call but that's another discussion)

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html (hunk 2 above)
Comment (menski) [reply broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: Any ideas how to visualize this? Maybe a plus icon in as first column?

**Comment by menski** [comment]
> However, I'm not sure if the debug exporter is the right place to put it in and who need this

I put it there cause then the configuration requirement is minimal. I often log records to debug the broker and noticed that the elasticsearch exporter is helpful to explain and discuss Zeebe. But as setting up elastic is some effort I thought this would be a nice addition. It also kind of provides a way to export the latest records as JSON over the `/records` endpoint, which is nice to share or attach the log to a issue. As it does not add any additional dependencies and is optional I thought it might be okay.

**Comment by npepinpe** [comment]
I think the idea is nice, but bundling everything into an exporter is strange; how about having a debug mode which is more than just an exporter, with the exporter being a component of it?

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (saig0) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporterHttpServer.java]: You're right. I didn't see the static reference.

File: broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (saig0) [reply broker-core/src/main/java/io/zeebe/broker/exporter/debug/DebugExporter.java]: Ok, make sense.

File: broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html (hunk 2 above)
Comment (saig0) [reply broker-core/src/main/resources/io/zeebe/broker/exporter/debug/index.html]: Yeah, this would be an option. Or change the mouse symbol.

**Comment by saig0** [comment]
I like that it is built as an exporter and that it is in the distro. But I would prefer an extra exporter because the HTTP is not related to the DEBUG one.

**Comment by menski** [comment]
I reverted the changes on the existing debug exporter but renamed it to `DebugLogExporter`. I then extracted the `DebugHttpExporter` to an own exporter. The `DebugHttpServer` now serves all resources including javascript and css. The `DebugHttpServer` and therefore the `DebugHttpExporter` will fail if the server cannot be started or a resource is not found.

I also made requested changes to the javascript logic and the cursor is now a pointer to indicate that the table rows can be expanded.

I updated the `zeebe.cfg.toml` to contain both exporters.

**Review by saig0 — APPROVED** [review]

**Pull request merged**